File: 백준/Gold/3055. 탈출/탈출.cc
#include <iostream>
#include <algorithm>
#include <queue>
#define MAX 52

using namespace std;

char arr[MAX][MAX];
int dis[MAX][MAX];
int dx[4] = {-1, 0, 1, 0}; // 상,우,하,좌
int dy[4] = {0, 1, 0, -1};
queue<pair<int, int>> waterQ;
queue<pair<int, int>> copyWaterQ;
queue<pair<int, int>> sQ;
queue<pair<int, int>> copy_sQ;
queue<pair<int, int>> emptyQ;
int r, c;

bool vaild(int x, int y){
    if (x >=1 && y >=1 && x <= r && y <= c) return true;
    else return false;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    bool check = false;
    cin >> r >> c;

    // 티떱숲 지도 초기화
    for (int i=1; i<=r; i++) {
        for (int j=1; j<=c; j++) {
            cin >> arr[i][j];

            if (arr[i][j] == '*')
                waterQ.push({i,j});
            else if (arr[i][j] == 'S')
                sQ.push({i,j});
        }
    }

    // cout << '\n';

    int cnt=1;

    while(true){

        // 고슴도치가 비버굴에 도착하는 경로
        while(!sQ.empty()){
            int cur_r = sQ.front().first;
            int cur_c = sQ.front().second;

            sQ.pop();

            if (arr[cur_r][cur_c] == '*') continue;
            else {
                for (int i=0; i<4; i++){
                    int nx = cur_r + dx[i];
                    int ny = cur_c + dy[i];

                    if (vaild(nx, ny) && arr[nx][ny] == '.'){
                        arr[nx][ny] = 'S';
                        dis[nx][ny] = dis[cur_r][cur_c] + 1;
                        copy_sQ.push({nx, ny});
                    }
                    else if (vaild(nx, ny) && arr[nx][ny] == 'D'){
                        check = true;
                        dis[nx][ny] = dis[cur_r][cur_c] + 1;
                        sQ = queue<pair<int, int>> ();
                        copy_sQ = queue<pair<int, int>> ();
                        break;
                    }
                }
            }
        } // sQ while문 종료
        // cout << "반복횟수: " << cnt << '\n';
        //     for (int i=1; i<=r; i++) {
        //         for (int j=1; j<=c; j++) {
        //             cout << arr[i][j];
        //         }
        //         cout << "\n";
        //     }
        //     cout << '\n';

        swap(sQ, copy_sQ);

        // 물이 상하좌우로 퍼지는 마법
        while(!waterQ.empty()){
            int cur_r = waterQ.front().first;
            int cur_c = waterQ.front().second;

            waterQ.pop();

            for(int i=0; i<4; i++){
                int nx = cur_r + dx[i];
                int ny = cur_c + dy[i];

                // 상하좌우에 물 범람하기
                if (vaild(nx, ny) && arr[nx][ny] == '.'){
                    arr[nx][ny] = '*';
                    copyWaterQ.push({nx, ny});
                }
                else if (vaild(nx, ny) && arr[nx][ny] == 'S'){
                    arr[nx][ny] = '*';
                    copyWaterQ.push({nx, ny});
                }
            }
        } // waterQ while문 종료
        
        swap(waterQ, copyWaterQ);

        // cout << "반복횟수: " << cnt << '\n';
        // for (int i=1; i<=r; i++) {
        //         for (int j=1; j<=c; j++) {
        //             cout << arr[i][j];
        //         }
        //         cout << "\n";
        // }
        // cout << '\n';
        cnt++;
        if (sQ.empty()) break;
    }

    if (check == false) cout << "KAKTUS\n";
    else {
        for (int i=1; i<=r; i++) {
            for (int j=1; j<=c; j++) {
                if (arr[i][j] == 'D'){
                    cout << dis[i][j] << '\n';
                }
            }
        }
    }

    return 0;
}

File: 백준/Gold/3055. 탈출/탈출.py
# 탈출

import sys
from collections import deque

MAX = 52

# 상하좌우
dx = [-1,1,0,0]
dy = [0,0,-1,1]

def valid(x,y):
     if 0<=x <r and 0<= y <c: return True
     return False

check = False
r,c = map(int, sys.stdin.readline().split())

# 티떱숲 지도 초기화
arr = []
waterq = deque()
copyWaterq = deque()
sq = deque()
copySq = deque()

for i in range(r):  
     arr.append(list(input()))

     for j in range(c):
          if arr[i][j] == '*':
               waterq.append((i,j))
          elif arr[i][j] == 'S':
               sq.append((i,j))

cnt =1
dis = [[0 for j in range(MAX)] for i in range(MAX)]

while True:
     # 고슴도치가 비버굴에 도착하는 경로
     while sq:
          x,y = sq.popleft()

          if arr[x][y] == '*':
               continue
          else:
               for i in range(4):
                    nx = x + dx[i]
                    ny = y + dy[i]

                    if valid(nx,ny) and arr[nx][ny] == '.':
                         arr[nx][ny] ='S'
                         dis[nx][ny] =dis[x][y] +1
                         copySq.append((nx,ny))
                    elif valid(nx,ny) and arr[nx][ny] == 'D':
                         check = True
                         dis[nx][ny] = dis[x][y]+1
                         sq = deque()
                         copySq = deque()
                         break
          
     sq, copySq = copySq, sq

     # 물이 상하좌우로 퍼지는 마법
     while waterq:
          x,y = waterq.popleft()

          for i in range(4):
               nx = x + dx[i]
               ny = y + dy[i]

               # 상하좌우에 물 범람하기
               if valid(nx,ny) and arr[nx][ny] == '.':
                    arr[nx][ny] = '*'
                    copyWaterq.append((nx,ny))
               elif valid(nx,ny) and arr[nx][ny] == 'S':
                    arr[nx][ny] = '*'
                    copyWaterq.append((nx,ny))
     
     waterq, copyWaterq = copyWaterq, waterq
     cnt +=1
     if len(sq) ==0: break

if check == False: print('KAKTUS')
else:
     for i in range(r):
          for j in range(c):
               if arr[i][j] == 'D':
                    print(dis[i][j])


